### PR TITLE
fix(migrations): fix generation of empty migrations

### PR DESCRIPTION
**Pull request by Co-Sic**
Migrations are still being created with `migrator.createMigration()` even if there is no schema diff. The current implementation of getSchemaDiff returns an Array with an empty string inside for an empty string as schema diff.

### CODE REVIEW

File: packages/migrations/src/Migrator.ts [hunk 1]
```diff
@@ -175,15 +175,7 @@ export class Migrator {
...
-    for (let i = lines.length - 1; i > 0; i--) {
-      if (lines[i]) {
-        break;
-      }
-
-      delete lines[i];
-    }
-
-    return lines;
+    return lines.filter(Boolean);
```
Comment (B4nan) [inline packages/migrations/src/Migrator.ts:178]: This code should stay as is, it does keep the blank lines in the middle of the dump on purpose, it is removing just those at the end. 

I think the issue you are talking about should be just about changing the for loop condition from `i > 0` to `i >= 0` so we also iterate on the first line.

File: tests/Migrator.test.ts [hunk 2]
```diff
@@ -133,8 +133,8 @@ describe('Migrator', () => {
 
   test('migration is skipped when no diff', async () => {
     const migrator = new Migrator(orm.em);
-    const getSchemaDiffMock = jest.spyOn<any, any>(Migrator.prototype, 'getSchemaDiff');
-    getSchemaDiffMock.mockResolvedValueOnce([]);
+    const schemaGeneratorMock = jest.spyOn<any, any>(SchemaGenerator.prototype, 'getUpdateSchemaSQL');
```
Comment (B4nan) [inline tests/Migrator.test.ts:136]: why did you change the test? this test should be about how `Migrator.getSchemaDiff()` works, not about `SchemaGenerator.getUpdateSchemaSQL()`.

if you want to test it on the `SchemaGenerator` level, adjust its tests, not `Migrator` ones, but as you were changing only Migrator code, I would expect the test to be same, and a new  test that would fail without your changes should be added.

**Review by B4nan — CHANGES_REQUESTED** [review]

File: packages/migrations/src/Migrator.ts (hunk 1 above)
Comment (Co-Sic) [reply packages/migrations/src/Migrator.ts]: Yes i confused break with continue. I changed it now, but i still need to change the delete operator, since it is not changing the length of the array. So for an empty string it will return [undefined], and the empty migration will still be generated. I replaced it with splice now.

File: tests/Migrator.test.ts (hunk 2 above)
Comment (Co-Sic) [reply tests/Migrator.test.ts]: I checked the the test file and there was not a single test on a private method and i am not sure myself how i would test a private method. That's why i put it in this test case. Can you point me to an example of a private method test, or if this is not possible or not desired, that is an alternative?

Comment (B4nan) [reply tests/Migrator.test.ts]: Test the behaviour, not implementation. You said you had issue with empty migration being generated - so in the new test, you could mock the schema generator to return the problematic dump (empty string in array) and see if the `createMigration` method will generate something or not. Be sure to verify that the test fails without the changes.

Comment (Co-Sic) [reply tests/Migrator.test.ts]: But if i am testing behavior my test case is already described in the test i changed originally: "migration is skipped when no diff". 
That's exactly the behavior that i want to test, or that is not working right now. Since that behavior depends on the correct functionality of `getSchemaDiff`, i mocked `SchemaGenerator.getUpdateSchemaSQL()` instead of `getSchemaDiff` to include the functionality of that function in the test. 
So the new test would just be a "deeper" version of the current test.

**Review by B4nan — CHANGES_REQUESTED** [review]
There are few tests that will need to be fixed, other than that this is LGTM 👍 

```
Summary of all failing tests
FAIL tests/Migrator.test.ts
  ● Migrator › generate js schema migration

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `Migrator generate js schema migration: migration-js-dump 1`

    - Snapshot  - 2
    + Received  + 0

    @@ -24,10 +24,8 @@
          "alter table `test2` drop foreign key `test2_foo___bar_foreign`;",
          "alter table `test2` drop index `test2_foo___bar_unique`;",
          "alter table `test2` drop index `test2_foo___bar_index`;",
          "alter table `test2` drop `foo___bar`;",
          "alter table `test2` drop `foo___baz`;",
    -     undefined,
    -     undefined,
        ],
        "fileName": "Migration20191013214813.js",
      }

      49 |     const migrator = orm.getMigrator();
      50 |     const migration = await migrator.createMigration();
    > 51 |     expect(migration).toMatchSnapshot('migration-js-dump');
         |                       ^
      52 |     orm.config.set('migrations', migrationsSettings); // Revert migration config changes
      53 |     await remove(process.cwd() + '/temp/migrations/' + migration.fileName);
      54 |   });

      at Object.<anonymous> (tests/Migrator.test.ts:51:23)

  ● Migrator › generate migration with custom name

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `Migrator generate migration with custom name: migration-dump 1`

    - Snapshot  - 2
    + Received  + 0

    @@ -21,10 +21,8 @@
          "alter table `test2` drop foreign key `test2_foo___bar_foreign`;",
          "alter table `test2` drop index `test2_foo___bar_unique`;",
          "alter table `test2` drop index `test2_foo___bar_index`;",
          "alter table `test2` drop `foo___bar`;",
          "alter table `test2` drop `foo___baz`;",
    -     undefined,
    -     undefined,
        ],
        "fileName": "migration-20191013214813.ts",
      }

      61 |     const migrator = orm.getMigrator();
      62 |     const migration = await migrator.createMigration();
    > 63 |     expect(migration).toMatchSnapshot('migration-dump');
         |                       ^
      64 |     const upMock = jest.spyOn(umzug.prototype, 'up');
      65 |     upMock.mockImplementation(() => void 0);
      66 |     const downMock = jest.spyOn(umzug.prototype, 'down');

      at Object.<anonymous> (tests/Migrator.test.ts:63:23)

  ● Migrator › generate schema migration

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `Migrator generate schema migration: migration-dump 1`

    - Snapshot  - 2
    + Received  + 0

    @@ -21,10 +21,8 @@
          "alter table `test2` drop foreign key `test2_foo___bar_foreign`;",
          "alter table `test2` drop index `test2_foo___bar_unique`;",
          "alter table `test2` drop index `test2_foo___bar_index`;",
          "alter table `test2` drop `foo___bar`;",
          "alter table `test2` drop `foo___baz`;",
    -     undefined,
    -     undefined,
        ],
        "fileName": "Migration20191013214813.ts",
      }

      83 |     const migrator = new Migrator(orm.em);
      84 |     const migration = await migrator.createMigration();
    > 85 |     expect(migration).toMatchSnapshot('migration-dump');
         |                       ^
      86 |     await remove(process.cwd() + '/temp/migrations/' + migration.fileName);
      87 |   });
      88 | 

      at Object.<anonymous> (tests/Migrator.test.ts:85:23)

  ● Migrator › generate initial migration

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `Migrator generate initial migration: initial-migration-dump 1`

    - Snapshot  - 2
    + Received  + 0

    @@ -331,10 +331,8 @@
          "alter table `user2_cars` add index `user2_cars_car2_name_car2_year_index`(`car2_name`, `car2_year`);",
          "",
          "alter table `author2` add index `custom_idx_name_123`(`name`);",
          "alter table `author2` add index `author2_name_age_index`(`name`, `age`);",
          "alter table `author2` add unique `author2_name_email_unique`(`name`, `email`);",
    -     undefined,
    -     undefined,
        ],
        "fileName": "Migration20191013214813.ts",
      }

      117 |     const migration1 = await migrator.createInitialMigration(undefined);
      118 |     expect(logMigrationMock).not.toBeCalledWith('Migration20191013214813.ts');
    > 119 |     expect(migration1).toMatchSnapshot('initial-migration-dump');
          |                        ^
      120 |     await remove(process.cwd() + '/temp/migrations/' + migration1.fileName);
      121 | 
      122 |     await orm.em.getKnex().schema.dropTableIfExists(orm.config.get('migrations').tableName!);

      at Object.<anonymous> (tests/Migrator.test.ts:119:24)

  ● Migrator › generate initial migration

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `Migrator generate initial migration: initial-migration-dump 2`

    - Snapshot  - 2
    + Received  + 0

    @@ -331,10 +331,8 @@
          "alter table `user2_cars` add index `user2_cars_car2_name_car2_year_index`(`car2_name`, `car2_year`);",
          "",
          "alter table `author2` add index `custom_idx_name_123`(`name`);",
          "alter table `author2` add index `author2_name_age_index`(`name`, `age`);",
          "alter table `author2` add unique `author2_name_email_unique`(`name`, `email`);",
    -     undefined,
    -     undefined,
        ],
        "fileName": "Migration20191013214813.ts",
      }

      123 |     const migration2 = await migrator.createInitialMigration(undefined);
      124 |     expect(logMigrationMock).toBeCalledWith('Migration20191013214813.ts');
    > 125 |     expect(migration2).toMatchSnapshot('initial-migration-dump');
          |                        ^
      126 |     await remove(process.cwd() + '/temp/migrations/' + migration2.fileName);
      127 |   });
      128 | 

      at Object.<anonymous> (tests/Migrator.test.ts:125:24)

FAIL tests/Migrator.postgres.test.ts
  ● Migrator (postgres) › generate js schema migration

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `Migrator (postgres) generate js schema migration: migration-js-dump 1`

    - Snapshot  - 2
    + Received  + 0

    @@ -16,10 +16,8 @@
      ",
        "diff": Array [
          "alter table \"book2\" drop column \"foo\";",
          "",
          "alter table \"test2\" drop column \"path\";",
    -     undefined,
    -     undefined,
        ],
        "fileName": "Migration20191013214813.js",
      }

      49 |     const migrator = orm.getMigrator();
      50 |     const migration = await migrator.createMigration();
    > 51 |     expect(migration).toMatchSnapshot('migration-js-dump');
         |                       ^
      52 |     orm.config.set('migrations', migrationsSettings); // Revert migration config changes
      53 |     await remove(process.cwd() + '/temp/migrations/' + migration.fileName);
      54 |   });

      at Object.<anonymous> (tests/Migrator.postgres.test.ts:51:23)

  ● Migrator (postgres) › generate migration with custom name

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `Migrator (postgres) generate migration with custom name: migration-dump 1`

    - Snapshot  - 2
    + Received  + 0

    @@ -13,10 +13,8 @@
      ",
        "diff": Array [
          "alter table \"book2\" drop column \"foo\";",
          "",
          "alter table \"test2\" drop column \"path\";",
    -     undefined,
    -     undefined,
        ],
        "fileName": "migration-20191013214813.ts",
      }

      61 |     const migrator = orm.getMigrator();
      62 |     const migration = await migrator.createMigration();
    > 63 |     expect(migration).toMatchSnapshot('migration-dump');
         |                       ^
      64 |     const upMock = jest.spyOn(umzug.prototype, 'up');
      65 |     upMock.mockImplementation(() => void 0);
      66 |     const downMock = jest.spyOn(umzug.prototype, 'down');

      at Object.<anonymous> (tests/Migrator.postgres.test.ts:63:23)

  ● Migrator (postgres) › generate schema migration

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `Migrator (postgres) generate schema migration: migration-dump 1`

    - Snapshot  - 2
    + Received  + 0

    @@ -13,10 +13,8 @@
      ",
        "diff": Array [
          "alter table \"book2\" drop column \"foo\";",
          "",
          "alter table \"test2\" drop column \"path\";",
    -     undefined,
    -     undefined,
        ],
        "fileName": "Migration20191013214813.ts",
      }

      83 |     const migrator = new Migrator(orm.em);
      84 |     const migration = await migrator.createMigration();
    > 85 |     expect(migration).toMatchSnapshot('migration-dump');
         |                       ^
      86 |     await remove(process.cwd() + '/temp/migrations/' + migration.fileName);
      87 |   });
      88 | 

      at Object.<anonymous> (tests/Migrator.postgres.test.ts:85:23)

  ● Migrator (postgres) › generate initial migration

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `Migrator (postgres) generate initial migration: initial-migration-dump 1`

    - Snapshot  - 2
    + Received  + 0

    @@ -193,10 +193,8 @@
          "alter table \"author2_following\" add constraint \"author2_following_author2_2_id_foreign\" foreign key (\"author2_2_id\") references \"author2\" (\"id\") on update cascade on delete cascade;",
          "",
          "create index \"custom_idx_name_123\" on \"author2\" (\"name\");",
          "create index \"author2_name_age_index\" on \"author2\" (\"name\", \"age\");",
          "alter table \"author2\" add constraint \"author2_name_email_unique\" unique (\"name\", \"email\");",
    -     undefined,
    -     undefined,
        ],
        "fileName": "Migration20191013214813.ts",
      }

      117 |     const migration1 = await migrator.createInitialMigration(undefined);
      118 |     expect(logMigrationMock).not.toBeCalledWith('Migration20191013214813.ts');
    > 119 |     expect(migration1).toMatchSnapshot('initial-migration-dump');
          |                        ^
      120 |     await remove(process.cwd() + '/temp/migrations/' + migration1.fileName);
      121 | 
      122 |     await orm.em.getKnex().schema.dropTableIfExists(orm.config.get('migrations').tableName!);

      at Object.<anonymous> (tests/Migrator.postgres.test.ts:119:24)

  ● Migrator (postgres) › generate initial migration

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `Migrator (postgres) generate initial migration: initial-migration-dump 2`

    - Snapshot  - 2
    + Received  + 0

    @@ -193,10 +193,8 @@
          "alter table \"author2_following\" add constraint \"author2_following_author2_2_id_foreign\" foreign key (\"author2_2_id\") references \"author2\" (\"id\") on update cascade on delete cascade;",
          "",
          "create index \"custom_idx_name_123\" on \"author2\" (\"name\");",
          "create index \"author2_name_age_index\" on \"author2\" (\"name\", \"age\");",
          "alter table \"author2\" add constraint \"author2_name_email_unique\" unique (\"name\", \"email\");",
    -     undefined,
    -     undefined,
        ],
        "fileName": "Migration20191013214813.ts",
      }

      123 |     const migration2 = await migrator.createInitialMigration(undefined);
      124 |     expect(logMigrationMock).toBeCalledWith('Migration20191013214813.ts');
    > 125 |     expect(migration2).toMatchSnapshot('initial-migration-dump');
          |                        ^
      126 |     await remove(process.cwd() + '/temp/migrations/' + migration2.fileName);
      127 |   });
      128 | 

      at Object.<anonymous> (tests/Migrator.postgres.test.ts:125:24)


Snapshot Summary
 › 10 snapshots failed from 2 test suites. Inspect your code changes or run `yarn test -u` to update them.

Test Suites: 2 failed, 137 passed, 139 total
Tests:       8 failed, 926 passed, 934 total
Snapshots:   10 failed, 82 passed, 92 total
Time:        269.402 s
Ran all test suites.
```

**Review by B4nan — APPROVED** [review]
